Initial commit of environ

elisarver opened the pull request:
This package manipulates a copy of the environment for later exec calls. It's got two really nice functions, Keep and Drop, that take lists of variable names/globbed prefix values.

Commit:
Initial commit of environ, a package to manipulate env variables for use by exec.

Commit:
update readme text

shairozan reviewed (approved):
I like the simplicity and the design in general. After our discussion about the use of the readlocker / writelocker helpers, I think it's good to go for now.

Commit:
created internal test for ()() operation check

- run every function calling locks
- confirm with a result that states search term for finding offender.

Commit:
created internal test for ()() operation check

- run every function calling locks
- confirm with a result that states search term for finding offender.

dpastoor commented on README.md:
```diff
@@ -1,2 +1,35 @@
+env.Keep("PATH", "SHELL", "GITHUB_*", "SSH_*")
+
+// Drop specific keys that might confound a process.
+env.Drop("GITHUB_SHA")
```
can drop also do wildcard matches? (this is the first file i'm reviewing so this will likely get answered later)

dpastoor replied:
answer - yes it does

dpastoor commented on environ.go:
```diff
@@ -0,0 +1,304 @@
+		var found bool
+
+		// Matching on wildcards
+		if strings.HasSuffix(keeperKey, "*") {
```
what about using one of the globbing packages? that would give us more globbing functionality around like what about if glob was `*_TOKEN` or `AWS_*_KEY`

elisarver replied:
Originally, I was going to search for a match on `"[-_]"` to limit it to complete matches. The real use case is applications with their own prefixes, not expansions on globs.

elisarver replied:
I don't know why you'd use `*_TOKEN` instead of the *exact* token you need to feed into the environment. I also think you'd just want `AWS_*` because the domain is more important they type of value (in this case `_KEY`)

dpastoor replied:
i'm thinking more about drop than keep - eg I currently have GHE_TOKEN, GITHUB_TOKEN, JIRA_API_TOKEN, and METWORX_API_TOKEN in my environ, so from a sanitization perspective the situation of, I want to match the global env a bit but want to just put some protection is, especially thinking in terms of this in general purpose settings. 

That said, its such a small thing to add if we ever have an actual demonstrated use case, this likely will cover the vast majority of cases, and worst case a higher level implementation could still just glob themselves and pass in the end results.

My one request would then be, can we update the readme to reflect its just a prefix glob (I will add suggestion)

elisarver replied:
I'm just going to switch to using a regexp and returning err if it doesn't compile.

dpastoor commented on environ.go:
```diff
@@ -0,0 +1,304 @@
+// AsSlice emits the contents of the Environ as a slice of string
+// with a "key=value" format.
+func (e *Environ) AsSlice() []string {
+	defer e.readLocker()()
```
i'm trying to wrap my head around how this is working. Is this immediately invoking e.readLocker() which then returns a function, but that function is deferred to execute. What is the reason for this over an explicit like

```
e.Lock()
defer e.Unlock()
....
```

elisarver replied:
The reason is often forgetting to defer the unlock, mostly. I have a test that would work for either style, and I can replace this "clever" defer with a less clever inlined pair in all locations.

elisarver replied:
this also prevents:

```
e.Lock()
defer e.RUnlock()
```

elisarver replied:
In fact, that was the problem I was trying to solve with that construct, because I forgot several of the defers or used the wrong unlock function before implementing the locker closure.

dpastoor reviewed (approved):
This is great - just some clarifying implementation questions but will be useful across a number of libs immediately!

Commit:
Add TOOL_* glob example to Drop()

dpastoor commented on README.md:
```diff
@@ -1,2 +1,35 @@
+- Concurrent safety.
+- Functions for reading the set in different forms. (slice, map)
+- Functions over the set of values (Len())
+- Globbing of keys to keep or drop whole collections of values.
```
I think this is what threw me off to the globbing commentary below, I read globbing of keys and was thinking full glob patterns until reading the implementation (is this supposed to call out something of different than the keep/drop functions noted above which does explicitly call that out?). 

Perhaps update this to say like suffix globbing of keys to... (and to that, is it a actually prefix or suffix - I would have thought it implied what was globbed, so prefix glob would be *foo and suffix glob would bee foo*)

dpastoor commented on environ_internal_test.go:
```diff
@@ -0,0 +1,52 @@
+	}
+}
+
+type fakeLocker struct {
```
just a comment, I haven't seen this fake pattern used to test a mutex before - very neat

elisarver replied:
This is why private interfaces are sometimes useful. Implementation of the interface is be able to test the locking is complete by replacing a real lock with this and checking lock count after running all locking functions.

dpastoor reviewed (commented):
Some more thought on this - one additional thought to raise for discussion that came up a while ago with pkgr implementation - the use of a map as the underlying datastructure given we can't keep the environment variable ordering controlled.  The situation I think is more boogie-man than actual concern, given that the real place of interest was more within the PATH variable not really "all environment variables possible". Is there any reason to think that losing that ordering could be problematic where we might need a less efficient but more controllable data structure. 

I believe the implementation we went with was just a list where we just did that slice modification around getting/setting.

I also think some of the pain we felt was more in testing - IIRC we didn't do any sorting on the way out so test results would come back different and that was annoying...

Commit:
use regular expressions to handle key comparisons

elisarver commented:
Committed a change that moves the matching to its own function, and uses regexp patterns to do matching. They are `^+pattern+$` anchored so they don't bleed, so you need to be precise that what you select with your pattern will match the whole variable name.
